src/tasks/wifi.c

```diff
@@ -259,15 +271,74 @@ static void process_rx_msgs(struct Serial* s)
                 case RX_BUFF_STATUS_OVERFLOW:
                         /* A message (possibly partial) awaits us */
                         process_ready_msg(s);
+			msg_handled = true;
                         goto rx_done;
                 }
         }
 
 rx_done:
         rx_buff_clear(rxb);
         state.rx_msgs.timeout = 0;
+	return msg_handled;
+}
+
+/**
+ * Polls all of our Serial devices to see if one of them has a message.
+ * Will process all serial devices in completely fair queueing mannerisms
+ * until all available messages have been processed.
+ */
+static void check_connections()
+{
+	bool msg_handled = false;
+	for (int i = 0; i < ARRAY_LEN(state.connections); ++i) {
+		struct Serial* serial = state.connections[i];
+		if (!serial)
+			continue;
+
+		/* Check if serial is closed.  If so, handle it */
+		if (!serial_is_connected(serial)) {
+			pr_info_int_msg(LOG_PFX "Closing external "
+					"connection: ", i);
+			state.connections[i] = NULL;
+			logger_sample_disable_callback(serial);
+			serial_destroy(serial);
+			continue;
+		}
+
+		msg_handled |= process_rx_msg(serial);
+	}
+
+	/*
+	 * Check if a message was handled.  If so then schedule again.
+	 * We do this to ensure we don't starve the event queue of resources
+	 * since this can take a while to do.
+	 */
+	if (msg_handled)
+		check_connections_cb(NULL);
 }
 
+static void process_new_connection(struct Serial *s)
+{
+	/* Find an empy connection slot.  If none, close it */
+	for (int i = 0; i < ARRAY_LEN(state.connections); ++i) {
+		if (state.connections[i])
+			continue;
+
+		/*
+		 * If here, found slot. Set the serial IOCTL handler here
+		 * b/c this is managing task
+		 */
+		pr_info_int_msg(LOG_PFX "New external connection: ", i);
+		serial_set_ioctl_cb(s, wifi_serial_ioctl);
+		state.connections[i] = s;
+		return;
+	}
+
+	/* If here, no slot found.  Close the Serial port */
+	pr_warning(LOG_PFX "Max external connections reached. Closing\r\n");
```

Comment:
The meat of the change is as follows:

* Serial devices can now close.
* All Serial read/write ops return number of characters read/written, or -1 as an error code.
* All read/write ops were evaluated and refactored to ensure their return codes were interpreted correctly
* Esp8266_driver is only responsible for creating Serial objects when a new connection arrives or is requested or closing serial objects when a connection disappears or is closed internally.
* WiFi task is now responsible for managing sockets and reaping their resources when they close.